### PR TITLE
Handle analytical mode in community_detection during parallel changes

### PR DESCRIPTION
**Pull request by Darych**
### Description

NB: PR taken over by @antepusic; connected with https://github.com/memgraph/memgraph/pull/1395

Since there are no ACID guarantees in analytical storage mode, the graph may get modified by parallel transactions while the user’s query is running. Query procedures and functions that return graph elements (nodes and relationships) should be robust to the deletion of graph elements, and this PR aims to fix that.

This PR contains changes to query modules that ensure the intended behavior:
* Query modules written with the C and C++ APIs check the existence of values that may contain graph elements before returning them (Python API query modules do this under the hood)

Merge commit message:

> Add proper handling of deleted return values for query procedures and functions ran in analytical mode

### Pull request type

- [x] Bugfix

### Related issues

Resolves https://github.com/memgraph/memgraph/issues/1036 (together with https://github.com/memgraph/memgraph/pull/1395)

######################################

#### Module/Algorithm
- [ ] Core algorithm/module implementation
- [ ] [Query module](https://memgraph.com/docs/memgraph/reference-guide/query-modules) implementation
- [ ] Unit tests
- [ ] End-to-end tests
- [ ] Code documentation
- [ ] README short description
- [ ] Documentation on [memgraph/docs](https://github.com/memgraph/docs)
- [ ] Update GQLALchemy signatures in [query builder](https://github.com/memgraph/gqlalchemy/blob/main/gqlalchemy/graph_algorithms/query_builder.py  ) using [query module signature generator](https://github.com/memgraph/gqlalchemy/blob/main/scripts/query_module_signature_generator.py)

######################################

### CODE REVIEW

File: cpp/community_detection_module/community_detection_online_module.cpp
Comment (as51340) [inline cpp/community_detection_module/community_detection_online_module.cpp:88]: Cache result of calling graph.GetNodeById, no need to call it twice

File: cpp/community_detection_module/community_detection_online_module.cpp
Comment (as51340) [inline cpp/community_detection_module/community_detection_online_module.cpp:186]: Same

File: cpp/community_detection_module/community_detection_online_module.cpp
Comment (as51340) [inline cpp/community_detection_module/community_detection_online_module.cpp:199]: What exceptions can be thrown in `GetNodeById`?

File: cpp/mg_utility/data_structures/graph_view.hpp
Comment (as51340) [inline cpp/mg_utility/data_structures/graph_view.hpp:111]: Memgraph ID is GID and what is Inner Node ID?

File: cpp/mg_utility/mg_utils.hpp
Comment (as51340) [inline cpp/mg_utility/mg_utils.hpp:123]:
```suggestion
  graph->SetIsTransactional(mgp::graph_is_transactional(memgraph_graph));
```

File: cpp/pagerank_module/pagerank_module.cpp
Comment (as51340) [inline cpp/pagerank_module/pagerank_module.cpp:1]: Why these 2 things got included?

File: cpp/betweenness_centrality_module/betweenness_centrality_module.cpp
Comment (as51340) [inline cpp/betweenness_centrality_module/betweenness_centrality_module.cpp:28]: These code is duplicates over all modules... It would be good to respect to respect DRY principle here and abstract this code into a method.

File: cpp/mg_utility/mg_graph.hpp
Comment (as51340) [inline cpp/mg_utility/mg_graph.hpp:335]:
```suggestion
    auto it = memgraph_to_inner_id_.find(memgraph_id); 
    if (it != memgraph_to_inner_id_.end()) {
        return *it;
     }
    if (IsTransactional()) {
        throw mg_exception::InvalidIDException();
    }
    return std::nullopt;
    }
```

**Review by as51340 — CHANGES_REQUESTED** [review]
I think good job in general, take care about duplication please :)

**Comment by as51340** [comment]
And you have build errors, please take care about that before requesting review again. For next time tip: it is better to request review after build errors are fixed since it gives reviewer more confidence that hey this is good and meaningful fix without negative impact on the rest of the system

File: cpp/betweenness_centrality_module/betweenness_centrality_module.cpp
Comment (antepusic) [reply cpp/betweenness_centrality_module/betweenness_centrality_module.cpp:28]: Will do

File: cpp/mg_utility/data_structures/graph_view.hpp
Comment (antepusic) [reply cpp/mg_utility/data_structures/graph_view.hpp:111]: The inner node ID is specific to the `GraphView` data structure used by a few existing query modules. For a graph with `n` nodes, their inner IDs are 0...`n - 1`.

File: cpp/pagerank_module/pagerank_module.cpp
Comment (antepusic) [reply cpp/pagerank_module/pagerank_module.cpp:1]: Will remove – I added them to help with replicating the issue.

File: cpp/katz_centrality_module/katz_centrality_online_module.cpp
Comment (Darych) [inline cpp/katz_centrality_module/katz_centrality_online_module.cpp:100]: Instead of calling `graph.GetNodeById` once again, can we use maybe_node here?

File: cpp/katz_centrality_module/katz_centrality_online_module.cpp
Comment (Darych) [inline cpp/katz_centrality_module/katz_centrality_online_module.cpp:96]: Since the function returns not optional Node, IMO it's better to name the variable as node and reuse it later.

File: cpp/pagerank_module/pagerank_online_module.cpp
Comment (Darych) [inline cpp/pagerank_module/pagerank_online_module.cpp:126]: The same comment as in `katz_centrality_online_module.cpp`

File: cpp/community_detection_module/community_detection_online_module.cpp
Comment (Darych) [reply cpp/community_detection_module/community_detection_online_module.cpp:199]: `NotFoundException`
https://github.com/memgraph/memgraph/blob/master/include/mgp.hpp#L1959

File: cpp/mg_utility/data_structures/graph_view.hpp
Comment (Darych) [reply cpp/mg_utility/data_structures/graph_view.hpp:111]: Do we still need this function in the interface?

File: cpp/mg_utility/mg_graph.hpp
Comment (antepusic) [reply cpp/mg_utility/mg_graph.hpp:335]: Done

File: cpp/mg_utility/data_structures/graph_view.hpp
Comment (antepusic) [reply cpp/mg_utility/data_structures/graph_view.hpp:111]: We do – it’s used by older query modules (and their tests) that rely on `GraphView`.

File: cpp/mg_utility/mg_utils.hpp
Comment (antepusic) [reply cpp/mg_utility/mg_utils.hpp:123]: Done

File: cpp/mg_utility/mg_graph.hpp
Comment (antepusic) [reply cpp/mg_utility/mg_graph.hpp]: Done

File: cpp/katz_centrality_module/katz_centrality_online_module.cpp
Comment (antepusic) [reply cpp/katz_centrality_module/katz_centrality_online_module.cpp:96]: Done

File: cpp/pagerank_module/pagerank_online_module.cpp
Comment (antepusic) [reply cpp/pagerank_module/pagerank_online_module.cpp:126]: Done

**Comment by Darych** [comment]
My initial idea was to make body of `GetInnerNodeIdOpt` a default implementation for `GetInnerNodeId`, so we will have just one implementation. I mentioned about this [HERE](https://github.com/memgraph/memgraph/issues/1036)
Currently. we have 2 implementations and users should think which one they need to use in their modules.
@antoniofilipovic what do you think?

**Review by Darych — COMMENTED** [review]
LGTM. I can't approve the PR since I'm author.

File: cpp/betweenness_centrality_module/betweenness_centrality_module.cpp
Comment (antepusic) [reply cpp/betweenness_centrality_module/betweenness_centrality_module.cpp]: Done

File: .github/workflows/test.yml
Comment (as51340) [inline .github/workflows/test.yml:10]: why?

**Review by as51340 — APPROVED** [review]

**Comment by antepusic** [comment]
@vpavicic As this PR is the MAGE side of https://github.com/memgraph/memgraph/pull/1395/, they are covered by the same docs PR (https://github.com/memgraph/documentation/pull/361) and should both be linked in the same release note.

File: .github/workflows/test.yml
Comment (antepusic) [reply .github/workflows/test.yml:10]: As v2.13 is not released yet, we’re running the checks on the current master (i.e. unofficial). After merging, this should be fixed in the "prepare mage for release" PR.

**Pull request merged**
